File: Numerical_Methods/Numerical Integration Techniques/Simpson’s 13 Rule/python_solutions/problem2.py
import numpy as np

def f(x):
    """Function to calculate f(x) = sin(x)"""
    return np.sin(x)

def simpsons_rule(a, b, n):
    """
    Implementation of Simpson's 1/3 Rule
    
    Parameters:
    a : float - Lower limit of integration
    b : float - Upper limit of integration
    n : int - Number of intervals (must be even)
    
    Returns:
    float - Approximate value of the integral
    """
    if n % 2 != 0:
        raise ValueError("Number of intervals must be even")
    
    h = (b - a) / n
    x = np.linspace(a, b, n+1)
    y = f(x)
    
    # Apply Simpson's 1/3 rule formula
    result = h/3 * (y[0] + y[-1] + 4*np.sum(y[1:-1:2]) + 2*np.sum(y[2:-1:2]))
    return result

def main():
    try:
        # Fixed values for this problem
        a = 0.0
        b = np.pi
        n = 6
        
        # Calculate using Simpson's 1/3 rule
        result = simpsons_rule(a, b, n)
        
        # Exact value of ∫sin(x) dx from 0 to π is 2
        exact_value = 2.0
        
        # Calculate error
        error = abs(result - exact_value)
        
        # Print results
        print("Simpson's 1/3 Rule Results:")
        print("---------------------------")
        print(f"Number of intervals (n): {n}")
        print(f"Approximate value: {result:.10f}")
        print(f"Exact value: {exact_value:.10f}")
        print(f"Absolute error: {error:.10f}")
        print(f"Relative error: {(error/exact_value)*100:.10f}%")
        
    except Exception as e:
        print(f"Error: {str(e)}")

if __name__ == "__main__":
    main() 

File: Numerical_Methods/Numerical Integration Techniques/Weddle’s Rule/java_solutions/Problem2.java
public class Problem2 {
    public static double f(double x) {
        return Math.sin(x);
    }

    public static double weddlesRule(double a, double b, int n) {
        if (n % 6 != 0) {
            throw new IllegalArgumentException("Number of intervals must be divisible by 6");
        }

        double h = (b - a) / n;
        double sum1 = f(a) + f(b);  // First and last terms
        double sum2 = 0, sum3 = 0, sum4 = 0, sum5 = 0;

        for (int i = 1; i < n; i++) {
            double x = a + i * h;
            if (i % 6 == 1) {
                sum2 += f(x);
            } else if (i % 6 == 2 || i % 6 == 4) {
                sum3 += f(x);
            } else if (i % 6 == 3) {
                sum4 += f(x);
            } else if (i % 6 == 5) {
                sum5 += f(x);
            }
        }

        sum2 *= 5;
        sum4 *= 6;
        sum5 *= 5;

        return (3 * h / 10) * (sum1 + sum2 + sum3 + sum4 + sum5);
    }

    public static void main(String[] args) {
        double a = 0;           // Lower limit
        double b = Math.PI/2;   // Upper limit
        int n = 6;             // Number of intervals

        try {
            double result = weddlesRule(a, b, n);
            double exactValue = -Math.cos(b) + Math.cos(a);  // Integral of sin(x) is -cos(x)
            double error = Math.abs(result - exactValue);

            System.out.printf("Approximate value using Weddle's Rule: %.6f%n", result);
            System.out.printf("Exact value: %.6f%n", exactValue);
            System.out.printf("Absolute error: %.6f%n", error);
        } catch (IllegalArgumentException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
} 